homeassistant/components/homematic.py

```diff
@@ -0,0 +1,528 @@
+        for dev in dev_descriptions:
+            key_dict[dev['ADDRESS'].split(':')[0]] = True
+        # Connect devices already created in HA to pyhomematic and
+        # add remaining devices to list
+        devices_not_created = []
+        for dev in key_dict:
+            try:
+                if dev in HOMEMATIC_DEVICES:
+                    for hm_element in HOMEMATIC_DEVICES[dev]:
+                        hm_element.link_homematic()
+                else:
+                    devices_not_created.append(dev)
+            # pylint: disable=broad-except
+            except Exception as err:
+                _LOGGER.error("Failed to setup device %s: %s", str(dev),
+                              str(err))
+        # If configuration allows autodetection of devices,
+        # all devices not configured are added.
+        if HOMEMATIC_AUTODETECT and devices_not_created:
+            for component_name, discovery_type in (
+                    ('switch', DISCOVER_SWITCHES),
+                    ('light', DISCOVER_LIGHTS),
+                    ('rollershutter', DISCOVER_ROLLERSHUTTER),
+                    ('binary_sensor', DISCOVER_BINARY_SENSORS),
+                    ('sensor', DISCOVER_SENSORS),
+                    ('thermostat', DISCOVER_THERMOSTATS)):
+                # Get all devices of a specific type
+                try:
+                    found_devices = _get_devices(discovery_type,
+                                                 devices_not_created)
+                # pylint: disable=broad-except
+                except Exception as err:
+                    _LOGGER.error("Failed generate opt %s with error '%s'",
+                                  component_name, str(err))
+
+                # When devices of this type are found
+                # they are setup in HA and an event is fired
+                if found_devices:
+                    try:
+                        component = get_component(component_name)
+                        config = {component.DOMAIN: found_devices}
+
+                        # Ensure component is loaded
+                        homeassistant.bootstrap.setup_component(
+                            _HM_DISCOVER_HASS,
+                            component.DOMAIN,
+                            config)
+
+                        # Fire discovery event
+                        _HM_DISCOVER_HASS.bus.fire(
+                            EVENT_PLATFORM_DISCOVERED, {
+                                ATTR_SERVICE: discovery_type,
+                                ATTR_DISCOVERED: {
+                                    ATTR_DISCOVER_DEVICES:
+                                    found_devices,
+                                    ATTR_DISCOVER_CONFIG: ''
+                                    }
+                                }
+                            )
+                    # pylint: disable=broad-except
+                    except Exception as err:
```

Comment:
There is no chance here that an exception will be raised. Discovery fires an event, actually setting up the platform is done in another thread.